include/seqan3/core/detail/pack_algorithm.hpp

```diff
@@ -0,0 +1,220 @@
+    //!\cond
+        requires std::invocable<wrapper_t, std::type_identity<args_t>...>
+    //!\endcond
+    static constexpr std::invoke_result_t<wrapper_t, std::type_identity<args_t>...> invoke(wrapper_t && wrapper)
+    {
+        return wrapper(std::type_identity<args_t>{}...);
+    }
+};
+
+//-----------------------------------------------------------------------------
+// all_of
+//-----------------------------------------------------------------------------
+
+/*!\brief Tests whether a given predicate evaluates to `true` for each element in a parameter pack.
+ * \ingroup core
+ *
+ * \tparam unary_predicate_t The function type, like function pointers, functors and lambdas;
+ *                           must model std::predicate expanded on each argument type.
+ * \tparam pack_t The parameter pack of the arguments (each argument type can be different).
+ *
+ * \param[in] fn The predicate to evaluate for every argument.
+ * \param[in] args The parameter pack.
+ *
+ * \details
+ *
+ * This function behaves like std::all_of but on parameter packs. The invocation(s) will be done without any loop.
+ *
+ * ### Example
+ *
+ * \include test/snippet/core/detail/all_of_in_pack.cpp
+ *
+ * ### Complexity
+ *
+ * Linear in the number of elements in the pack.
+ *
+ * \attention Opposed to the std::all_of the argument order is changed, such that the first argument is the unary
+ * predicate to invoke on each argument followed by the arguments.
+ *
+ * \sa https://en.cppreference.com/w/cpp/language/parameter_pack
+ */
+template <typename unary_predicate_t, typename ...pack_t>
+[[nodiscard]] constexpr bool all_of(unary_predicate_t && fn, pack_t && ...args)
+//!\cond
+    requires (std::predicate<unary_predicate_t, pack_t> && ...)
+//!\endcond
+{
+    return (fn(std::forward<pack_t>(args)) && ...);
+}
+
+/*!\brief Tests whether a given predicate evaluates to `true` for each type in a type list.
+ * \ingroup core
+ *
+ * \tparam list_t A type list; must model seqan3::detail::type_list_specialisation or seqan3::tuple_like
+ * \tparam unary_predicate_t The function type, like function pointers, functors and lambdas;
+ *                           must model std::Predicate expanded on each argument type wrapped in std::type_identity.
+ *
+ * \param[in] fn The predicate called for every type in the type list.
+ *
+ * \details
+ *
+ * This function behaves like std::all_of but on parameter packs. The types will be wrapped into std::type_identity and
+ * passed as arguments to seqan3::detail::all_of. The invocation(s) will be done without any loop.
+ *
+ * ### Example
+ *
+ * \include test/snippet/core/detail/all_of_in_type_list.cpp
+ *
+ * ### Complexity
+ *
+ * Linear in the number of types in the list. Requires linear number of template instantiations.
+ */
+template <typename list_t, typename unary_predicate_t>
+[[nodiscard]] constexpr bool all_of(unary_predicate_t && fn)
+//!\cond
+    requires type_list_specialisation<list_t> || tuple_like<list_t>
```

Comment:
I am happy to remove the tuple based version and it is true, that if one wants to pass a tuple one can use the transformation trait to make it a type_list. This makes the interface explicit and also the expectations for this function.

Reply:
> I think type_list_specialisation is a weird / clunky name and we should consider a rename. I think before it was called TypeList that makes this a tough one to rename.

I know. @h-2  and I had a discussion about it. And while we both weren't really satisfied with the name we couldn't find a better solution. So if you have better suggestion that is always welcome. Put it on the discussion board maybe.

> if we also want to allow tuple_like, we should introduce a new concept that extracts the requires clause.

No I don't think we want this. As you said there is already a transformation available for tuple to type_list and that is expressive enough. The tuple should not be used to just store the types. It would be possible to use std::apply on the tuple and call the algorithm on the expanded arguments there.

Reply:
okay good!